iceoryx_posh/include/iceoryx_posh/popo/sample.hpp

```diff
@@ -81,69 +134,51 @@ class Sample
     operator bool() const;
 
     ///
-    /// @brief allocation Access to the memory chunk loaned to the sample.
-    /// @return
+    /// @brief allocation Access to the encapsulated type loaned to the sample.
+    /// @return a pointer to the encapsulated type.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
     T* get() noexcept;
 
     ///
-    /// @brief allocation Read-only access to the memory chunk loaned to the sample.
-    /// @return
+    /// @brief allocation Read-only access to the encapsulated type loaned to the sample.
+    /// @return a const pointer to the encapsulated type.
     ///
     const T* get() const noexcept;
 
     ///
     /// @brief header Retrieve the header of the underlying memory chunk loaned to the sample.
     /// @return The ChunkHeader of the underlying memory chunk.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
     mepoo::ChunkHeader* getHeader() noexcept;
 
+    ///
+    /// @brief header Retrieve the header of the underlying memory chunk loaned to the sample.
+    /// @return The const ChunkHeader of the underlying memory chunk.
+    ///
+    const mepoo::ChunkHeader* getHeader() const noexcept;
+
     ///
     /// @brief publish Publish the sample via the publisher from which it was loaned and automatically
     /// release ownership to it.
+    /// @details Only available for non-const type T.
     ///
+    template <typename S = T, typename = std::enable_if_t<std::is_same<S, T>::value && !std::is_const<S>::value, S>>
```

Comment:
We require the rule: Always prefer `static_assert` over `enable_if`.
I think here we can again use `static_assert` to provide a nice error message for the user.
Maybe the rule could be: only use `enable_if` when you would like to do some template specialization.

Reply:
in this case the compiler would probably choose the non-const method in a non-const environment, which triggers the static_assert and makes it awkward to use the correct method

Reply:
Forget this. The publisher expects a sample and therefore `publish` cannot be in a base class.

Reply:
I'm a bit torn about the `publish` method. On the one hand, it's nice syntactical sugar. On the other hand, it complicates stuff and with the refactoring it will also show up in the auto-completion of the IDE. I have an idea which might help with the auto-completion, but I would rather spent my time on more important stuff. How about removing this feature for 1.0 and bringing it back afterwards?

@budrus @ithier what do you think?

Reply:
@elBoberido What is the issue with the auto-completion. It suggests publish for the `Sample` on the Subscriber side? 
It seems to be a cool feature. We do not have much experience how people use the API but I already see that we use this feature when I look at [spontaneously written code](https://github.com/eclipse-iceoryx/iceoryx/issues/581#issuecomment-785789697). Your bad IDE would not be a good reason for me to remove this feature ;-) Or are there also other aspects?

Reply:
@budrus it's not really an issue. If the method is selected it is marked as unavailable. I also checked what happens when this method is removed and the code looks a bit awkward with so many `publish` words in a row. So let's keep it.